src/transformers/models/maskformer/modeling_maskformer.py

```diff
@@ -0,0 +1,1638 @@
+        x,
+        size=(h, w),
+        mode=mode,
+        align_corners=False,
+    )
+    return upsampled
+
+
+class MaskFormerModel(PreTrainedModel):
+    config_class = MaskFormerConfig
+    base_model_prefix = "model"
+    main_input_name = "pixel_values"
+
+    def __init__(self, config: MaskFormerConfig):
+        super().__init__(config)
+        self.pixel_level_module = MaskFormerPixelLevelModule(config)
+        self.transformer_module = MaskFormerTransformerModule(
+            in_features=self.pixel_level_module.backbone.get_outputs_shape()[-1], config=config
+        )
+        self.segmentation_module = MaskFormerSegmentationModule(config)
+        self.matcher = MaskFormerHungarianMatcher(
+            cost_class=1.0, cost_dice=config.dice_weight, cost_mask=config.mask_weight
+        )
+
+        losses = ["labels", "masks"]
+        # TODO I don't like too aggressive abbreviations in the naming
+        # what is 'ce'? It's cross entropy but it's not clear
+        # ask if we should follow a more exhaustive naming convention
+        self.weight_dict: Dict[str, float] = {
+            "loss_ce": config.ce_weight,
+            "loss_mask": config.mask_weight,
+            "loss_dice": config.dice_weight,
+        }
+
+        self.criterion = MaskFormerLoss(
+            config.num_classes,
+            matcher=self.matcher,
+            weight_dict=self.weight_dict,
+            eos_coef=config.no_object_weight,
+            losses=losses,
+        )
+
+    def forward(self, pixel_values: Tensor, targets: Optional[Dict[str, Tensor]] = None) -> MaskFormerOutput:
+        image_features, pixel_embeddings = self.pixel_level_module(pixel_values)
+        queries = self.transformer_module(image_features)
+        outputs: Dict[str, Tensor] = self.segmentation_module(queries, pixel_embeddings)
+
+        loss_dict: Dict[str, Tensor] = {}
+        loss: Tensor = None
+        if targets:
+            loss_dict = self.get_loss_dict(outputs, targets)
+            loss = self.get_loss(loss_dict)
+        else:
+            # upsample the masks to match the inputs' spatial dimension
+            outputs["pred_masks"] = upsample_like(outputs["pred_masks"], pixel_values)
+
+        return MaskFormerOutput(**outputs, loss_dict=loss_dict, loss=loss)
+
+    def get_loss_dict(self, outputs: Dict[str, Tensor], targets: Dict[str, Tensor]) -> Dict[str, Tensor]:
+        loss_dict: Dict[str, Tensor] = self.criterion(outputs, targets)
+        # weight each loss by `self.weight_dict[<LOSS_NAME>]`
+        weighted_loss_dict: Dict[str, Tensor] = {k: v * self.weight_dict[k] for k, v in loss_dict.items()}
+        return weighted_loss_dict
+
+    def get_loss(self, loss_dict: Dict[str, Tensor]) -> Tensor:
+        # probably an awkward way to reduce it
+        return torch.tensor(list(loss_dict.values()), dtype=torch.float).sum()
+
+
+@dataclass
+class MaskFormerForSemanticSegmentationOutput(MaskFormerOutput):
```

Comment:
We usually don't use inheritance on model outputs types, to avoid having a sequence to follow. Just redefine the list of fields here.

Reply:
Indeed, and please place it at the top of the modeling file.